src/Middleware/RetryMiddleware.php

```diff
@@ -86,12 +95,27 @@ public function __invoke(Call $call, array $options)
         }
 
         return $nextHandler($call, $options)->then(null, function ($e) use ($call, $options) {
-            if (!$e instanceof ApiException) {
-                throw $e;
-            }
-
-            if (!in_array($e->getStatus(), $this->retrySettings->getRetryableCodes())) {
-                throw $e;
+            $customRetryFunc = $this->retrySettings->getRequestRetryFunction();
+            // If we have a custom retry function, and it returns false
+            // we throw the exception, otherwise we continue.
+            if (!is_null($customRetryFunc)) {
+                if (!$customRetryFunc($e, $options, $this->currentAttempts)) {
+                    throw $e;
+                }
+
+                // since the $customRetryFunction returned true,
+                // retry will take place
+            } else {
+                // This is the default retry behaviour, i.e. we don't retry an ApiException
+                // and for other exception types, we only retry when the error code is in
+                // the list of retryable error codes.
+                if (!$e instanceof ApiException) {
+                    throw $e;
+                }
+
+                if (!in_array($e->getStatus(), $this->retrySettings->getRetryableCodes())) {
+                    throw $e;
+                }
```

Comment:
Just a suggestion, but instead of `$customRetryFunc` we should just call it `retryFunction`, and have a private method `getDefaultRetryFunction()` with the following:

```php
return function (Exception $e, int $retryAttempt): bool {
    // This is the default retry behaviour, i.e. we don't retry an ApiException
    // and for other exception types, we only retry when the error code is in
    // the list of retryable error codes.
    if (!$e instanceof ApiException) {
        throw $e;
    }

    if (!in_array($e->getStatus(), $this->retrySettings->getRetryableCodes())) {
        throw $e;
    }
};
```

Reply:
Thanks for the suggestion. Incorporated.